File: js/store.js
// Persistence layer — everything lives in localStorage, no backend.
// Keys are versioned so future schema changes can migrate cleanly.

const K = {
  settings: 'lid.settings.v1',
  progress: 'lid.progress.v1',
  stats: 'lid.stats.v1',
  exams: 'lid.exams.v1',
};

function read(key, fallback) {
  try {
    const raw = localStorage.getItem(key);
    return raw ? JSON.parse(raw) : fallback;
  } catch {
    return fallback;
  }
}

function write(key, value) {
  localStorage.setItem(key, JSON.stringify(value));
}

export const store = {
  get settings() {
    const defaultExam = new Date(Date.now() + 14 * 24 * 3600 * 1000).toISOString().slice(0, 10);
    return { theme: 'auto', ...read(K.settings, { state: null, lang: 'en', examDate: defaultExam }) };
  },
  set settings(s) {
    write(K.settings, s);
  },
  patchSettings(patch) {
    this.settings = { ...this.settings, ...patch };
  },

  // progress: { [questionId]: {box, due, seen, right, wrong} }
  get progress() {
    return read(K.progress, {});
  },
  set progress(p) {
    write(K.progress, p);
  },

  // stats: { days: { 'YYYY-MM-DD': {answered, correct} } }
  get stats() {
    return read(K.stats, { days: {} });
  },
  set stats(s) {
    write(K.stats, s);
  },

  get exams() {
    return read(K.exams, []);
  },
  set exams(e) {
    write(K.exams, e);
  },

  recordAnswer(qid, correct, now = Date.now()) {
    const day = new Date(now).toISOString().slice(0, 10);
    const stats = this.stats;
    const d = stats.days[day] || { answered: 0, correct: 0 };
    d.answered += 1;
    if (correct) d.correct += 1;
    stats.days[day] = d;
    this.stats = stats;
  },

  answeredToday(now = Date.now()) {
    const day = new Date(now).toISOString().slice(0, 10);
    return (this.stats.days[day] || { answered: 0 }).answered;
  },

  streak(now = Date.now()) {
    const days = this.stats.days;
    let n = 0;
    const d = new Date(now);
    for (;;) {
      const key = d.toISOString().slice(0, 10);
      if ((days[key] || {}).answered > 0) {
        n += 1;
        d.setDate(d.getDate() - 1);
      } else {
        // today without answers yet doesn't break the streak
        if (n === 0 && key === new Date(now).toISOString().slice(0, 10)) {
          d.setDate(d.getDate() - 1);
          continue;
        }
        break;
      }
    }
    return n;
  },

  exportAll() {
    return JSON.stringify({
      version: 1,
      exportedAt: new Date().toISOString(),
      settings: this.settings,
      progress: this.progress,
      stats: this.stats,
      exams: this.exams,
    });
  },

  importAll(json) {
    const data = JSON.parse(json);
    if (!data || data.version !== 1) throw new Error('Unbekanntes Format');
    if (data.settings) this.settings = data.settings;
    if (data.progress) this.progress = data.progress;
    if (data.stats) this.stats = data.stats;
    if (data.exams) this.exams = data.exams;
  },

  resetProgress() {
    localStorage.removeItem(K.progress);
    localStorage.removeItem(K.stats);
    localStorage.removeItem(K.exams);
  },
};

File: flutter_app/lib/store.dart
// Persistence — SharedPreferences with the SAME keys and JSON shapes as the
// PWA's localStorage (app/composables/useStore.ts), so exports from the web
// app import losslessly here and vice versa.
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'models.dart';

class _K {
  static const settings = 'lid.settings.v1';
  static const progress = 'lid.progress.v1';
  static const stats = 'lid.stats.v1';
  static const exams = 'lid.exams.v1';
  static const marked = 'lid.marked.v1';
  static const lastSaved = 'lid.lastSaved.v1';
  // Set on every launch while the app is sold paid-upfront. If the app later
  // switches to free + IAP, this flag grandfathers early buyers.
  static const paidUpfront = 'lid.paidUpfront.v1';
}

class Store extends ChangeNotifier {
  Store(this._prefs) {
    _prefs.setBool(_K.paidUpfront, true);
  }

  final SharedPreferences _prefs;

  static Future<Store> open() async => Store(await SharedPreferences.getInstance());

  T _read<T>(String key, T Function(dynamic) parse, T fallback) {
    final raw = _prefs.getString(key);
    if (raw == null) return fallback;
    try {
      return parse(jsonDecode(raw));
    } catch (_) {
      return fallback;
    }
  }

  void _write(String key, Object? value) {
    _prefs.setString(key, jsonEncode(value));
    _prefs.setString(_K.lastSaved, jsonEncode(DateTime.now().millisecondsSinceEpoch));
    notifyListeners();
  }

  // ---- settings ----

  Settings getSettings() => _read(
        _K.settings,
        (j) => Settings.fromJson(j as Map<String, dynamic>),
        Settings(examDate: Settings.defaultExamDate()),
      );

  void patchSettings({String? state, String? lang, String? examDate, String? theme}) {
    final s = getSettings();
    if (state != null) s.state = state;
    if (lang != null) s.lang = lang;
    if (examDate != null) s.examDate = examDate;
    if (theme != null) s.theme = theme;
    _write(_K.settings, s.toJson());
  }

  // ---- progress ----

  Map<String, ProgressEntry> getProgress() => _read(
        _K.progress,
        (j) => (j as Map<String, dynamic>)
            .map((k, v) => MapEntry(k, ProgressEntry.fromJson(v as Map<String, dynamic>))),
        <String, ProgressEntry>{},
      );

  void setProgress(Map<String, ProgressEntry> p) =>
      _write(_K.progress, p.map((k, v) => MapEntry(k, v.toJson())));

  // ---- daily stats / streak ----

  Map<String, dynamic> _getStatsRaw() => _read(
        _K.stats,
        (j) => (j as Map<String, dynamic>),
        <String, dynamic>{'days': <String, dynamic>{}},
      );

  // UTC like the PWA's `toISOString().slice(0, 10)` — keys must match exactly
  // or imported streaks/day counts split across two dates near midnight.
  static String _dayKey(DateTime d) => d.toUtc().toIso8601String().substring(0, 10);

  void recordAnswer(bool correct) {
    final stats = _getStatsRaw();
    final days = (stats['days'] as Map<String, dynamic>? ?? {});
    final day = _dayKey(DateTime.now());
    final d = (days[day] as Map<String, dynamic>? ?? {'answered': 0, 'correct': 0});
    d['answered'] = (d['answered'] as num).toInt() + 1;
    if (correct) d['correct'] = (d['correct'] as num).toInt() + 1;
    days[day] = d;
    stats['days'] = days;
    _write(_K.stats, stats);
  }

  int answeredToday() {
    final days = (_getStatsRaw()['days'] as Map<String, dynamic>? ?? {});
    final d = days[_dayKey(DateTime.now())] as Map<String, dynamic>?;
    return ((d?['answered'] as num?) ?? 0).toInt();
  }

  int streak() {
    final days = (_getStatsRaw()['days'] as Map<String, dynamic>? ?? {});
    int answered(String key) =>
        (((days[key] as Map<String, dynamic>?)?['answered'] as num?) ?? 0).toInt();
    var n = 0;
    var d = DateTime.now();
    // today may still be untouched; a streak kept alive until yesterday counts
    if (answered(_dayKey(d)) == 0) d = d.subtract(const Duration(days: 1));
    while (answered(_dayKey(d)) > 0) {
      n += 1;
      d = d.subtract(const Duration(days: 1));
    }
    return n;
  }

  // ---- exams ----

  List<ExamRecord> getExams() => _read(
        _K.exams,
        (j) => (j as List).map((e) => ExamRecord.fromJson(e as Map<String, dynamic>)).toList(),
        <ExamRecord>[],
      );

  void addExam(ExamRecord e) =>
      _write(_K.exams, [...getExams().map((x) => x.toJson()), e.toJson()]);

  // ---- manually bookmarked ("difficult") questions ----

  Set<String> getMarked() => _read(
        _K.marked,
        (j) => (j as List).cast<String>().toSet(),
        <String>{},
      );

  bool toggleMarked(String id) {
    final s = getMarked();
    final nowMarked = !s.contains(id);
    if (nowMarked) {
      s.add(id);
    } else {
      s.remove(id);
    }
    _write(_K.marked, s.toList());
    return nowMarked;
  }

  // ---- export / import (format identical to the PWA, version 1) ----

  String exportAll() => jsonEncode({
        'version': 1,
        'exportedAt': DateTime.now().toUtc().toIso8601String(),
        'settings': getSettings().toJson(),
        'progress': getProgress().map((k, v) => MapEntry(k, v.toJson())),
        'stats': _getStatsRaw(),
        'exams': getExams().map((e) => e.toJson()).toList(),
        'marked': getMarked().toList(),
      });

  void importAll(String json) {
    final data = jsonDecode(json);
    if (data is! Map<String, dynamic> || data['version'] != 1) {
      throw const FormatException('Unbekanntes Format');
    }
    if (data['settings'] != null) _write(_K.settings, data['settings']);
    if (data['progress'] != null) _write(_K.progress, data['progress']);
    if (data['stats'] != null) _write(_K.stats, data['stats']);
    if (data['exams'] != null) _write(_K.exams, data['exams']);
    if (data['marked'] != null) _write(_K.marked, data['marked']);
  }

  void resetProgress() {
    _prefs.remove(_K.progress);
    _prefs.remove(_K.stats);
    _prefs.remove(_K.exams);
    _prefs.remove(_K.marked);
    notifyListeners();
  }

  ({int seen, int? lastSaved}) storageHealth() {
    final seen = getProgress().values.where((e) => e.seen > 0).length;
    final lastSaved = _read<int?>(_K.lastSaved, (j) => (j as num).toInt(), null);
    return (seen: seen, lastSaved: lastSaved);
  }
}
